### PR TITLE
using LDAP users in UI tests

### PR DESCRIPTION
**Pull request by individual-it**
## Description
pseudo user creation in LDAP 

## Motivation and Context
we would like to run tests against a LDAP backend but not going throw the problems of creating the users/groups in PHP. So when running LDAP tests (TEST_LDAP env. variable set to "true") we assume the users/groups already exist in the LDAP server

So this would be a way to run at least some acceptance tests on LDAP with very minimal changes. See changes on the LDAP side https://github.com/owncloud/user_ldap/pull/173

## How Has This Been Tested?
run tests in drone https://github.com/owncloud/user_ldap/pull/173

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Review by phil-davis — COMMENTED** [review]
Looks like a good/flexible thing.
Having an "LDAP" tag on some scenarios gives the implication (not true) that the scenario might be just for LDAP. I am not sure how to get around that - people just need to understand what tags are used for, and that there is an implied "empty tag" that is "run all scenarios"

**Comment by individual-it** [comment]
@ownclouders rebase

**Comment by ownclouders** [comment]
Hey! I'm [GitMate.io](https://gitmate.io)! This pull request is being rebased automatically. Please **DO NOT** push while rebase is in progress or your changes would be lost permanently :warning:

**Comment by ownclouders** [comment]
Automated rebase with [GitMate.io](https://gitmate.io) was successful! :tada:

**Comment by individual-it** [comment]
cleaned stuff up, please review

File: tests/ui/features/bootstrap/BasicStructure.php [hunk 1]
```diff
@@ -185,8 +185,14 @@ public function theseUsersExist($doNotInitialize, TableNode $table) {
...
 	private function createUser(
 		$user, $password, $displayName = null, $email = null, $initialize = true,
-		$method = "api"
+		$method = null
 	) {
+		if ($method === null && getenv("TEST_EXTERNAL_USER_BACKENDS") === "true") {
+			//guess yourself
+			$method = "ldap";
+		} else {
+			$method = "api";
```
Comment (phil-davis) [inline tests/ui/features/bootstrap/BasicStructure.php]: If ``$method`` is passed in here as ``occ`` (for example) then this code will go down this "else" and set (override) ``$method`` to be ``api``

I don't see how method ``occ`` can be used?

File: tests/ui/features/bootstrap/BasicStructure.php [hunk 2]
```diff
@@ -217,6 +223,9 @@ private function createUser(
 					);
 				}
 				break;
+			case "ldap":
+				echo "creating LDAP users is not implemented, so assume they exists\n";
```
Comment (phil-davis) [inline tests/ui/features/bootstrap/BasicStructure.php]: s/they exists/they exist

File: tests/ui/features/bootstrap/BasicStructure.php [hunk 3]
```diff
@@ -274,7 +283,13 @@ public function regularGroupsExist() {
 	 * @return void
 	 * @throws Exception
 	 */
-	private function createGroup($group, $method = "api") {
+	private function createGroup($group, $method = null) {
+		if ($method === null && getenv("TEST_EXTERNAL_USER_BACKENDS") === "true") {
+			//guess yourself
+			$method = "ldap";
+		} else {
+			$method = "api";
```
Comment (phil-davis) [inline tests/ui/features/bootstrap/BasicStructure.php]: Same here - how to use method ``occ`` ?

File: tests/ui/features/bootstrap/BasicStructure.php [hunk 4]
```diff
@@ -299,6 +314,9 @@ private function createGroup($group, $method = "api") {
 					);
 				}
 				break;
+			case "ldap":
+				echo "creating LDAP groups is not implemented, so assume they exists\n";
```
Comment (phil-davis) [inline tests/ui/features/bootstrap/BasicStructure.php]: s/they exists/they exist

File: tests/ui/features/bootstrap/BasicStructure.php [hunk 5]
```diff
@@ -327,7 +345,13 @@ public function aRegularUserIsInARegularGroup() {
 	 * @return void
 	 * @throws Exception
 	 */
-	public function theUserIsInTheGroup($user, $group, $method = "api") {
+	public function theUserIsInTheGroup($user, $group, $method = null) {
+		if ($method === null && getenv("TEST_EXTERNAL_USER_BACKENDS") === "true") {
+			//guess yourself
+			$method = "ldap";
+		} else {
+			$method = "api";
```
Comment (phil-davis) [inline tests/ui/features/bootstrap/BasicStructure.php]: Same here - how to use method ``occ`` ?

**Review by phil-davis — COMMENTED** [review]
It will work as it is, but there are a few comments about the logic to handle the ``occ`` method.

**Comment by individual-it** [comment]
@phil-davis fixed the mentioned issues

**Review by phil-davis — APPROVED** [review]

**Comment by individual-it** [comment]
backport in  #30349

**Pull request merged**